src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,12 +225,118 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+            // character, a filter was selected, etc. If so, those items will be added as part of the refresh.
+            //
+            // The reason of adopting this approach is we want to minimize typing delays. There are two ways user might perceive a delay in typing.
+            // First, they could see a delay between typing a character and completion list being displayed if they want to examine the items available.
+            // Second, they might be typing continuously w/o paying attention to completion list, and simply expect the completion to do the "right thing"
+            // when a commit char is typed (e.g. commit "cancellationToken" when typing 'can$TAB$'). However, the commit could be delayed if completion is
+            // still waiting on the computation of all available items, which manifests as UI delays and in worst case timeouts in commit which results in
+            // unexpected behavior (e.g. typing 'can$TAB$' results in a 250ms UI freeze and still ends up with "can" instead of "cancellationToken".)
+            //
+            // This approach would ensure the computation of #2 will not be the cause of such delays, with the obvious trade off of potentially not providing
+            // expanded items until later (or never) in a completion session even if the feature is enabled. Note that in most cases we'd expect task #2 to finish
+            // in time and complete result would be available from the start of the session. However, even in the case only partial result is returned at the start,
+            // we still believe this is acceptable given how critical perf is in typing scenario.
+            // Additionally, expanded items are usually considered complementary. The need for them only rise occasionally (it's rare when users need to add imports,)
+            // and when they are needed, our hypothesis is because of their more intrusive nature (adding an import to the document) users would more likely to
+            // contemplate such action thus typing slower before commit and/or spending more time examining the list, which give us some opportunities
+            // to still provide those items later before they are truly required.            
+
+            if (!options.ShouldShowItemsFromUnimportNamspaces())
+            {
+                // No need to trigger expanded providers at all if the feature is disabled, just trigger core providers and return;
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, list, triggerLocation);
+                return context;
+            }
+
+            if (!session.TextView.Options.GetOptionValue(DefaultOptions.ResponsiveCompletionOptionId))
```

Comment:
```suggestion
            }
            else if (!session.TextView.Options.GetOptionValue(DefaultOptions.ResponsiveCompletionOptionId))
```

Reply:
I remember being asked in reviews to do the opposite in the past. Do we have a guideline on when should we use separate ifs or if...else if...else

Reply:
it's entirely related to the code.  for things like early bailing out, i have no problem with if... if... if... etc.

But this is a case where there are three effective alternate approaches we're taking, with a lot of similarity.  it's basically almsot like:

```
choice A: ...
choice B: ...
choice C: ...
```

In that case, my preference is to put them on equal footing and to be clear there is no bleeding of data or control flow between any of them.

this counts as a nit here, but i think it woudl really help.  esp. if we're going to break this into separate helpers.